### ISSUE TITLE
Sounds have their own life

### ISSUE THREAD

**Issue by thetredev** [CloudMracek/BWHorror#3]
I looked over the code and tried a couple things to fix the weird sound issue you're having (weird frequencies instead of foot steps are played).

I couldn't fix the issue, but I've made a few interesting discoveries (Void Linux, x64, using the Mednaffe emulator):

## Scenario 1
You don't play any sound in `gameInit()`:
```c
//int ch = play_sample(steps_addr, steps_sr);
```

This leads to the following:
- A weird cracking sound is being played at the start.
- The *yay* sound becomes the *steps* sound.
- Whenever you catch a ghost, *steps* is played (without any reverb). And it's played correct. No weird frequencies whatsoever
- But: no actual foot step sounds while walking, though.

## Scenario 2
You play *yay* in `gameInit()`:
```
int ch = play_sample(yay_addr, yay_sr); // after yay has been initialized
```

This leads to the following:
- A weird cracking sound is being played at the start instead of *yay*.
- **The foot steps are actually playing while walking!**
- But: they're played only once and you have to stop and start moving to hear them again (which is probably a completely different issue I guess).
- Also: A different kind of weird (and high) frequencies are played seemingly randomly OR after a specific/random amount time during which no other sound is playing (maybe even only after playing *steps* at least once, don't know for sure).
- Also: Same as Scenario 1, *steps* is played whenever you catch a ghost without any reverb.

## Scenario 3
You play *victory* in `gameInit()`:
```
int ch = play_sample(victory_addr, victory_sr); // after victory has been initialized
```

This leads to Scenario 2 again.

## Conclusion
My guess is that either your mechanism for loading the sounds (`VAG_Header` struct + `upload_sample()`) or your mechanism for playing the sounds is not implemented as it should be. I'll test more and keep this issue updated when I can.

**Comment by CloudMracek** [comment]
Yeah I'm doing some questionable stuff with the channel numbers. So they might be offset by one on the console.

**Comment by thetredev** [comment]
Sounds plausible.

**Comment by CloudMracek** [comment]
What emulator are you using so I can reproduce the error without burning through a thousand discs.
Can't reproduce it on duckstation or pcsx-redux

**Comment by thetredev** [comment]
[Mednaffe](https://github.com/AmatCoder/mednaffe). Should work on Windows too.

**Comment by thetredev** [comment]
Don't know if the BIOSes are included, I use these ones: https://gist.github.com/juanbrujo/cf55d223ad01927a48f9ebac9f50bdee

**Comment by gazaai** [comment]
> What emulator are you using so I can reproduce the error without burning through a thousand discs. Can't reproduce it on duckstation or pcsx-redux

I can reproduce this using Duckstation (0.1-5437-g06d6447e (master)) and using BIOS - SCPH-1001 [NTSC-U/C] from here:
https://www.psxdev.net/forum/viewtopic.php?t=56

Hope that helps

**Comment by gazaai** [comment]
So I've taken a quick look at the code for this and found something interesting.
If you switch out the following line within `gameInit()`:
```C
int ch = play_sample(steps_addr, steps_sr);
```

for these two instead-
```C
upload_sample(&steps_vag[1], SWAP_ENDIAN(steps_vag->size));
play_sample(steps_addr, steps_sr);
```

The reverb footstep repeats now as it should without playing Yay! at the end and stopping.
However I'm not sure why uploading the same sample twice would help.

I've also just realised that the weird/high frequencies only seem to happen when I use the PAL BIOS I don't get that issue with the NTSC one.

UPDATE-
Ignore this, the issue appears to be more related to the `freeSpuRAM()` method as described in my post below.

**Comment by gazaai** [comment]
I've had another go at this and have found some interesting results with Duckstation (BIOS -> Fast Boot disabled).
So with a fresh clone of the repo.

### PAL BIOS
A high pitched sound will play instead of the footstep sound. Very loud and annoying, all the other sounds appear to work fine.

### NTSC BIOS
The sound almost works however the foot steps don't repeat and they finish with the "Yay!" sound.

### Possible Fix?
Updating the `freeSpuRAM` method from-
```C
void freeSpuRAM(void)
{
	SpuWrite((const uint32_t *)NULL, 0x1010 + (((13 << 11) * 4) * 2)) + 64;	
}
```
To-
```C
void freeSpuRAM(void)
{
	SpuSetTransferMode(SPU_TRANSFER_BY_DMA);
	SpuSetTransferStartAddr(ALLOC_START_ADDR);
	SpuWrite((const uint32_t *)NULL, ALLOC_START_ADDR + (((13 << 11) * 4) * 2));	
	SpuIsTransferCompleted(SPU_TRANSFER_WAIT);
}
```
This seems to solve all my sound issues on both PAL and NTSC BIOS however I don't have real hardware to test this on so cant test it on that.
Also I am new to PSX Dev so I may be doing something incorrectly in this updated method.
I was unsure why there is a `+ 64` at the end of the original line
`SpuWrite((const uint32_t *)NULL, 0x1010 + (((13 << 11) * 4) * 2)) + 64;`
so I have removed this and also switched out `0x1010` for `ALLOC_START_ADDR` since its already defined.

Maybe @UNSTOP4BLE can chime in on if this is correct?

**Comment by thetredev** [comment]
Nice find!!

**Comment by UNSTOP4BLE** [comment]
I added that + 64 because thats what other homebrews i modified had 
#define ALLOC_START_ADDR (BUFFER_START_ADDR + (CHUNK_SIZE_MAX * 2) + 64)

**Comment by UNSTOP4BLE** [comment]
Im also pretty new to psx dev so im not sure what the + 64 does

**Comment by gazaai** [comment]
OK thanks for clarifying I think in this case it's not needed as the + 64 was outside the brackets so wasn't actually doing anything. Does anyone have a real PS1 to test this on?

**Comment by UNSTOP4BLE** [comment]
I got a pal ps1, i can test later

**Comment by thetredev** [comment]
@gazaai does your fix in `freeSpuRAM()` still need to upload the sound sample twice or is this unrelated?

**Comment by gazaai** [comment]
> @gazaai does your fix in `freeSpuRAM()` still need to upload the sound sample twice or is this unrelated?

No that's no longer required if you just update the freeSpuRAM like above

**Comment by thetredev** [comment]
Good to know, thanks!

**Comment by UNSTOP4BLE** [comment]
Update on a pal ps1, it plays the walk sound once and it doesnt do a yay after it

**Comment by thetredev** [comment]
Well AFAIK, PS games always had a separate release for PAL and NTSC each, anyways. So I guess we could just make this horror game a PAL game (which might still work perfectly fine on NTSC) and call it a day. @CloudMracek what do you think? Do you have a PAL PS? It should work for you first and foremost imo.

**Comment by CloudMracek** [comment]
Yes, I'm on PAL. This will likely be the best solution.

**Comment by thetredev** [comment]
@gazaai time for a PR then! :D

**Comment by gazaai** [comment]
@UNSTOP4BLE Thanks for doing the test, can you confirm that the sound repeats whilst holding down triangle and walking. Or do you mean that the sound only plays once and then cuts out even though your still walking?

**Comment by UNSTOP4BLE** [comment]
it only plays once and then cuts out

**Comment by gazaai** [comment]
> it only plays once and then cuts out

OK, would you mind trying one more thing please. I think we are getting close now.

Within `main.c` could you move the following line-
```C
int ch = play_sample(steps_addr, steps_sr);
```

To the very bottom of `gameInit()` just under `enableReverbOnChannels(1 << 0);`

I think it might be because we are playing the footstep sound before all the other sounds are loaded.
Hopefully that works :)

**Comment by UNSTOP4BLE** [comment]
Alright i will try that later

**Comment by UNSTOP4BLE** [comment]
same issue
